File: golang/signature/signature.go
package signature

import (
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"math/big"
	"net/url"
	"sort"
	"strings"
	"time"
)

func sortQueryString(queryString string) string {

	params, _ := url.ParseQuery(queryString)

	var paramKeys []string
	for paramKey := range params {
		paramKeys = append(paramKeys, paramKey)
	}
	sort.Strings(paramKeys)

	var sortedQueryParams []string
	for _, paramKey := range paramKeys {
		sortedQueryParams = append(sortedQueryParams, fmt.Sprintf("%s=%s", paramKey, params[paramKey][0]))
	}
	return strings.Join(sortedQueryParams, "&")
}

func Sha256(data string) string {
	h := sha256.New()
	h.Write([]byte(data))
	hashBytes := h.Sum(nil)

	return hex.EncodeToString(hashBytes)
}

func GenerateNonce(length int) (string, error) {
	const letters = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"
	nonce := make([]byte, length)
	for i := 0; i < length; i++ {
		num, err := rand.Int(rand.Reader, big.NewInt(int64(len(letters))))
		if err != nil {
			return "", err
		}
		nonce[i] = letters[num.Int64()]
	}

	return "n_" + string(nonce), nil
}

func HmacSha256(key []byte, data []byte) string {
	h := hmac.New(sha256.New, key)
	h.Write(data)
	hmacBytes := h.Sum(nil)

	return hex.EncodeToString(hmacBytes)
}

func GenerateSignature(requestMethod string, path string, parameters string) (string, int64, string, error) {
	nonce, err := GenerateNonce(40)
	if err != nil {
		return "", 0, "", err
	}
	timestamp := time.Now().Unix()
	randStr := fmt.Sprintf(
		"debank-api\n%s\n%d",
		nonce,
		timestamp,
	)
	randStrHash := Sha256(randStr)

	requestParams := fmt.Sprintf(
		"%s\n%s\n%s",
		strings.ToUpper(requestMethod),
		strings.ToLower(path),
		sortQueryString(strings.ToLower(parameters)),
	)
	requestParamsHash := Sha256(requestParams)

	signature := HmacSha256(
		[]byte(randStrHash),
		[]byte(requestParamsHash),
	)
	return nonce, timestamp, signature, nil
}

File: js/signature.js
const crypto = require("crypto");

function sortQueryString(queryString) {
  const params = new URLSearchParams(queryString);

  const sortedParams = Array.from(params).sort((a, b) =>
    a[0].localeCompare(b[0])
  );
  const sortedQueryString = new URLSearchParams(sortedParams).toString();

  return sortedQueryString;
}

function sha256(data) {
  return crypto.createHash("sha256").update(data);
}

function encodeText(text) {
  return new TextEncoder().encode(text);
}

function generateNonce(length = 32) {
  const characters =
    "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
  let nonce = "";
  for (let i = 0; i < length; i++) {
    nonce += characters.charAt(crypto.randomInt(0, characters.length));
  }
  return "n_" + nonce;
}

function hmacSha256(key, data) {
  return crypto.createHmac("sha256", key).update(data).digest("hex");
}

function generateSignature(requestMethod, path, parameters) {
  const nonce = generateNonce();
  const timestamp = Math.floor(Date.now() / 1000);
  const randStr = `debank-api\n${nonce}\n${timestamp}`;
  const randStrHash = encodeText(sha256(randStr).digest("hex"));
  const requestParams = `${requestMethod.toUpperCase()}\n${path.toLowerCase()}\n${sortQueryString(
    parameters.toLowerCase()
  )}`;
  const requestParamsHash = encodeText(sha256(requestParams).digest("hex"));

  const signature = hmacSha256(randStrHash, requestParamsHash);

  return {
    nonce,
    timestamp,
    signature,
  };
}

module.exports = { generateSignature };
